src/compose/volumes.ts

```diff
@@ -0,0 +1,231 @@
+
+		try {
+			return await this.createFromPath({ name, appId }, legacyPath);
+		} catch (e) {
+			this.logger.logSystemMessage(
+				`Warning: could not migrate legacy /data volume: ${e.message}`,
+				{ error: e },
+				'Volume migration error',
+			);
+		}
+	}
+
+	// oldPath must be a path inside /mnt/data
+	public async createFromPath(
+		opts: VolumeCreateOpts,
+		oldPath: string,
+	): Promise<void> {
+		const volume = await this.create(opts);
+		const handle = volume.dockerVolume;
+
+		// Convert the path to be of the same mountpoint so that rename can work
+		const volumePath = path.join(
+			constants.rootMountPoint,
+			'mnt/data',
+			...handle.Mountpoint.split(path.sep).slice(3),
+		);
+		await safeRename(oldPath, volumePath);
+	}
+
+	public async remove({ name, appId }: VolumeNameOpts) {
+		this.logger.logSystemEvent(LogTypes.removeVolume, { volume: { name } });
+		try {
+			await this.docker
+				.getVolume(Volumes.generateVolumeName({ name, appId }))
+				.remove();
+		} catch (e) {
+			this.logger.logSystemEvent(LogTypes.removeVolumeError, {
+				volume: { name, appId },
+				error: e,
+			});
+		}
+	}
+
+	public isEqualConfig(current: VolumeConfig, target: VolumeConfig): boolean {
+		const currentConfig = (_.mapKeys(current, (_v, k) =>
+			_.camelCase(k),
+		) as unknown) as ComposeVolume['config'];
+		const targetConfig = (_.mapKeys(target, (_v, k) =>
+			_.camelCase(k),
+		) as unknown) as ComposeVolume['config'];
+
+		const currentOpts = currentConfig.driverOpts || {};
+		const targetOpts = targetConfig.driverOpts || {};
+
+		const currentLabels = currentConfig.labels || {};
+		const targetLabels = targetConfig.labels || {};
+
+		return (
+			_.isEqual(currentOpts, targetOpts) &&
+			_.isEqual(currentLabels, targetLabels)
+		);
+	}
+
+	private static format(volume: Dockerode.VolumeInspectInfo): ComposeVolume {
+		const match = volume.Name.match(/^([0-9]+)_(.+)$/);
+		if (match == null) {
+			throw new Error('Malformed volume name in Volume.format');
+		}
+		const appId = checkInt(match[1]);
+		const name = match[2];
+
+		return {
+			name,
+			// We know this cast is fine due to the regex
+			appId: appId as number,
+			config: {
+				labels: _.omit(
+					ComposeUtils.normalizeLabels(volume.Labels),
+					_.keys(constants.defaultVolumeLabels),
+				),
+				driverOpts: volume.Options,
+			},
+			dockerVolume: volume,
+		};
+	}
+
+	private async listWithBothLabels(): Promise<Dockerode.VolumeInspectInfo[]> {
+		// We have to cast the listVolumes call from any[] to any below, until the
+		// relevant PR: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32383
+		// is merged and released
+		const [legacyResponse, currentResponse]: [
+			Dockerode.VolumeInfoList,
+			Dockerode.VolumeInfoList
+		] = await Promise.all([
+			this.docker.listVolumes({
+				filters: { label: ['io.resin.supervised'] },
+			}) as Promise<any>,
+			this.docker.listVolumes({
+				filters: { label: ['io.balena.supervised'] },
+			}) as Promise<any>,
```

Comment:
Any reason this can't be `Dockerode.VolumeInfoList` here?

Reply:
It can't be because the dockerode typings currently say that it returns `Promise<any[]>` which is not compatible with `Dockerode.VolumeInfoList`